Extract model crate

luckysori opened the pull request:
Continuing the efforts started with #1207, which was inspired by #226.

The extracted `model` crate still has some suspicious dependencies such as `sqlx`, which we should strive to remove (in the future).

We also extract a `bdk-ext` crate, which is mostly used for tests.

TODO:

- [x] Fix `sqlx-data.json`.
- [x] Separate `bdk-ext` extraction into independent commit.
- [x] Ensure that there are no unused dependencies in `daemon`.
- [x] Update `bors.toml`.

thomaseizinger commented on payout-curve/src/lib.rs:
```diff
@@ -0,0 +1,827 @@
+use anyhow::Context;
```
You seem to have attempted another crate here already? :)

thomaseizinger commented on bdk-ext/Cargo.toml:
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "bdk-ext"
```
Can this specialise as `bdk-test` perhaps?

thomaseizinger commented on model/Cargo.toml:
```diff
@@ -0,0 +1,37 @@
+ndarray_einsum_beta = "0.7.0"
+num = "0.4.0"
+parse-display = "0.5.3"
+reqwest = { version = "0.11", default-features = false }
```
Why?

luckysori replied:
I was unhappy about this too. We use the `Url` type. I wanted to depend on the `url` crate, but I wasn't sure if it would result in annoying incompatibilities between crates. What do you think?

thomaseizinger replied:
I think you can depend on `url`, `reqwest` just re-exports it in its latest version.

thomaseizinger commented on model/Cargo.toml:
```diff
@@ -0,0 +1,37 @@
+serde = { version = "1", features = ["derive"] }
+serde_json = "1"
+serde_with = { version = "1", features = ["macros"] }
+sqlx = { version = "0.5", features = ["sqlite", "runtime-tokio-rustls"] }
```
Couple of open questions regarding this:

- Should we trim the feature list here? Does that negatively (poisoned cache) or positively (less code) influence compile times?
- Can/Should we perhaps get rid of it entirely by mapping on the way out of the DB layer?

thomaseizinger commented on model/src/cfd.rs:
```diff
@@ -1737,13 +1815,13 @@ pub enum Completed<P, E> {
     },
 }
 
-impl<P, E> xtra::Message for Completed<P, E>
```
I think we may be able to get rid of this enum if we instead have separate commands for these:

- `setup_contract_failed`
- etc

That was the original plan at least. Certainly, we should no longer be sending this as a message so this can probably go away. Would be nice to do these cleanups in a patch at the front of the PR:

- git reset --soft HEAD^ to remove the commit but keep the changes
- git stash
- make the small patch
- git stash pop

thomaseizinger commented on model/src/lib.rs:
```diff
@@ -29,13 +24,27 @@ use std::ops::Sub;
+/// - The auto-rollover time-window (taker)
+pub const SETTLEMENT_INTERVAL: time::Duration = time::Duration::hours(24);
+
+pub const N_PAYOUTS: usize = 200;
```
This is an application decision IMO and should maybe stay in `daemon`?

luckysori replied:
Yeah, you're right. I'll return it to `daemon` and add a similar constant in the `model` tests.

thomaseizinger commented on model/src/cfd.rs:
```diff
@@ -1474,7 +1552,7 @@ impl Dlc {
     pub fn close_transaction(
         &self,
-        proposal: &crate::model::cfd::SettlementProposal,
+        proposal: &crate::cfd::SettlementProposal,
```
As a folllow-up, I think we should flatten this crate and then extract more meaningful modules. This `cfd` submodule is annoying me 🤣

thomaseizinger commented on model/src/cfd.rs:
```diff
@@ -2982,7 +3058,9 @@ mod tests {
 
     pub fn dummy_identity() -> Identity {
-        Identity::new(RandomSeed::default().derive_identity().0)
+        Identity::new(x25519_dalek::PublicKey::from(
+            *b"hello world, oh what a beautiful",
+        ))
```
Stuff like this would be nice in a separate patch. Then the last commit could all just be "move files".

thomaseizinger commented on model/src/lib.rs:
```diff
@@ -484,87 +493,6 @@ pub struct WalletInfo {
-#[derive(
-    Debug, Clone, Copy, SerializeDisplay, DeserializeFromStr, PartialEq, Eq, Hash, PartialOrd, Ord,
-)]
-pub struct BitMexPriceEventId {
```
This could be moved to the olivia module first as well to split the diff.

luckysori committed:
Remove unused xtra::Message impl

luckysori committed:
Do not depend on daemon::seed in model tests

A hard-coded `Identity` serves the same purpose as a random one in
these tests.

luckysori committed:
Move BitMexPriceEventId to daemon::olivia

The `BitMexPriceEventId` is part of Olivia's API.

luckysori commented on daemon/src/db.rs:
```diff
@@ -389,29 +389,34 @@ mod tests {
+        )
+    }
 
+    #[async_trait::async_trait]
```
Separate commit?

thomaseizinger replied:
I think we could also just solve this differently by making it a free function :)

luckysori replied:
Done. I wanted to see if I could keep a similar API with a trait, but needing `async_trait` is just stupid.

luckysori commented on daemon/src/oracle.rs:
```diff
@@ -344,15 +340,30 @@ impl xtra::Actor for Actor {
+        self.0.id
+    }
+}
+
```
Separate commit?

thomaseizinger replied:
It is motivated by creating `olivia::Attestation` right? Could move the entire creation of this dedicated model into a separate commit. We are starting to gold-plate here though :D

luckysori commented on model/src/cfd.rs:
```diff
@@ -3089,4 +3168,19 @@ mod tests {
+        let pk = bitcoin::PublicKey::new(secp256k1::PublicKey::from_secret_key(SECP256K1, &sk));
+
+        (sk, pk)
+    }
```
Separate commit?

luckysori commented on model/src/olivia.rs:
```diff
@@ -10,10 +10,18 @@ use time::macros::format_description;
+    pub price: u64,
+    pub scalars: Vec<SecretKey>,
+}
+
```
Separate commit?

thomaseizinger replied:
Perhaps yeah. Ties in with what I commented above. Up to you!

thomaseizinger commented on daemon/Cargo.toml:
```diff
@@ -16,10 +16,8 @@ hex = "0.4"
 itertools = "0.10"
 maia = "0.1.0"
+model = { path = "../model" }
 nalgebra = { version = "0.30", default-features = false, features = ["std"] }
```
What about `nalgebra`? Can this not be moved?

thomaseizinger commented on daemon/prepare_db.sh:
```diff
@@ -15,4 +15,4 @@ trap 'rm -f $TEMPDB' EXIT
 
 # prepare the sqlx-data.json rust mappings
-DATABASE_URL=sqlite:./$DAEMON_DIR/$TEMPDB SQLX_OFFLINE=true cargo sqlx prepare -- --bin taker
+DATABASE_URL=sqlite:./$DAEMON_DIR/$TEMPDB SQLX_OFFLINE=true cargo sqlx prepare
```
Nit: Separate commit? :)

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -1,20 +1,20 @@
 use anyhow::Result;
 use futures::future::BoxFuture;
 use futures::FutureExt;
+use model;
```
That should be unnecessary when it is a dedicated crate.

luckysori committed:
Extract test impl for Cfd into free functions

To allow us to separate the tests from the `Cfd` type in a future
patch.

luckysori committed:
Fix prepare_db.sh script

The crate `daemon` no longer has more than one binary (or any binary
for that matter), so we don't need to use any extra flags with `cargo
sqlx prepare`.

luckysori committed:
Move {Setup, Rollover}Params to model module

They are obviously part of the model, since they are the input to
their respective protocols.

luckysori committed:
Remove unused dependencies

Revealed by running `cargo udeps`.

luckysori committed:
Extract bdk-ext crate

luckysori committed:
Move sqlx macro to model module

Despite its purpose, it is currently only used in model-related
modules.

luckysori committed:
Move SETTLEMENT_INTERVAL constant to model module

luckysori committed:
Redefine N_PAYOUTS in model tests

The N_PAYOUTS used in the model tests doesn't have to match what we
use in the application (i.e. daemon).

luckysori committed:
Extract model crate

The extracted `model` crate still has some suspicious dependencies
such as `sqlx`, which we should strive to remove (in the future).

luckysori committed:
Add xtra-bitmex-price-feed to bors.toml

luckysori committed:
Further differentiate between olivia and oracle

The `olivia::{Attestation, Announcement}` is used for deserializing
and modelling. The `oracle::Attestation` is just a wrapper which
allows us to send around attestations between different
`xtra::Actor`s.

thomaseizinger commented on bors.toml:
```diff
@@ -11,6 +11,7 @@ status = [
   "clippy (xtras)",
   "clippy (bdk-ext)",
   "clippy (model)",
+  "clippy (xtra-bitmex-price-feed)",
```
Ups! Thank you :)
